### PR TITLE
Allow wildcards in hostmask

### PR DESCRIPTION
**Pull request by brunnre8**
According to https://modern.ircdocs.horse/#wildcard-expressions
masks should support "*" and "?" wildcards.
Within TL this only impacts the /ignore functionality.

The reasoning for doing this is to ignore say GuestNNNN!*@* with
guest*!*@* and be done with it if someone spams a gateway.

### CODE REVIEW

File: src/helper.js [hunk 1]
```diff
@@ -314,8 +315,19 @@ function parseHostmask(hostmask) {
...
+		compareWithWildcard(a.nick, b.nick) &&
+		compareWithWildcard(a.ident, b.ident) &&
+		compareWithWildcard(a.hostname, b.hostname)
 	);
 }
+
+function compareWithWildcard(a, b) {
+	a = a.toLowerCase();
+	b = b.toLowerCase();
+	const wildcars_split = a.split("*").map((sub) => {
```
Comment (itsjohncs) [inline src/helper.js:327]: Looks like this doesn't support escaping. The horse docs say `*` and `?` should be escapable with a `\`.

But TBH I'm not sure how important of a feature it is to support escaping so I wouldn't be too distraught if you just don't want to. Seems like it shouldn't be too difficult to support though.

File: src/helper.js [hunk 2]
```diff
@@ -314,8 +315,19 @@ function parseHostmask(hostmask) {
...
+		const parts = sub.split("?").map((p) => _.escapeRegExp(p));
+		return parts.join(".");
+	});
+	const re = new RegExp(wildcars_split.join(".*"));
```
Comment (itsjohncs) [inline src/helper.js:331]: If you split out the part of this function that generates the regex into its own function, you could write tests for that function that would be a bit easier to verify the correctness of.

If you pass `"i"` as a flag to `RegExp` it'll do a case insensitive comparison. Since we're using the regex engine now anyways we might as well let it handle that part of it.

**Review by itsjohncs — COMMENTED** [review]
Had a few minor thoughts, nothing blocking or particularly important though.

File: src/helper.js (hunk 1 above)
Comment (brunnre8) [reply src/helper.js:327]: Yeah... thought about it but then dismissed it as being probably overkill... after all who uses a wildcard in a nick / ident / host.

But while we are at it we might just do it properly don't we?

File: src/helper.js (hunk 2 above)
Comment (brunnre8) [reply src/helper.js:331]: ack, will do. Thanks

File: src/helper.js (hunk 1 above)
Comment (brunnre8) [reply src/helper.js:327]: FTR: the spec is very much unclear how you'd then escape the escape char.
Considering that nicks can contain the \ char but not the wildcards we refrain from implementing the escaping.

File: src/helper.js (hunk 2 above)
Comment (brunnre8) [reply src/helper.js]: I didn't split the function but added the regex modifier (and a huge comment).
Considering that there are already extensive tests for the comparison function with quite a bunch of edge case scenarios I don't see a benefit of testing for the raw regex, we already know that we get the desired result. Hope that's ok.

File: test/tests/hostmask.js [hunk 3]
```diff
@@ -60,3 +67,38 @@ describe("Hostmask", function () {
...
 });
+
+describe("compareWithWildcard", function () {
+	const goodPairs = [
+		["asdf", "asdf"],
+		["a?df*", "asdf"],
+		["*asdf*", "asdf"],
+		["*asdf", "asdf"],
+		["asd?", "asdf"],
+		["asd?*", "asdf"],
```
Comment (itsjohncs) [inline test/tests/hostmask.js:78]: Could you add these cases?

```js
["AsDf", "asdf"],
["a??f", "asdf"],
["a*", "asdf"],
["*f", "asdf"],
["*s*", "asdf"],
```

If `String.split()` worked like it did in Python, the last 4 cases would've failed. Thought I'd found a bug at first!

File: test/tests/hostmask.js [hunk 4]
```diff
@@ -60,3 +67,38 @@ describe("Hostmask", function () {
...
+		});
+	}
+
+	const badPairs = [
+		["asdf", "fdsa"],
+		["a?df*", "adfg"],
+		["?", ""],
+		["?asdf", "asdf"],
+		["?*", ""],
+		["*?*", ""],
```
Comment (itsjohncs) [inline test/tests/hostmask.js:95]: Could you add these cases?

```js
["asd", "asdf"],
["sdf", "asdf"],
["sd", "asdf"],
["", "asdf"],
```

They currently fail. You'll want to add `^` and `$` to the regex to make sure it matches the whole string.

**Review by itsjohncs — CHANGES_REQUESTED** [review]
I think it's fine as one function and not handling escaping sounds fine as well 👍.

I did find one bug that we ought to fix though. At least I think it's a bug. I don't think the mask `asd` is meant to match `asdf`.

File: test/tests/hostmask.js (hunk 4 above)
Comment (brunnre8) [reply test/tests/hostmask.js:95]: Right, nice catch! Thanks

File: test/tests/hostmask.js (hunk 3 above)
Comment (brunnre8) [reply test/tests/hostmask.js:78]: what do you mean?
```python
>>> "*".join("*s*".split("*"))
'*s*'
```

Comment (itsjohncs) [reply test/tests/hostmask.js:78]: Oh! I just got mixed up.

`split` without arguments will collapse empty splits but apparently not if you give a delimiter explicitly.

**Comment by brunnre8** [comment]
PTAL, added the tests you suggested and fixed the code

**Review by itsjohncs — APPROVED** [review]
Thanks! LGTM

**Review by MaxLeiter — APPROVED** [review]
